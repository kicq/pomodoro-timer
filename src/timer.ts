import { Ref, ref } from 'vue'
import countdown from '@/assets/countdown.wav'

export type State = 'running' | 'stopping' | 'suspended' | 'prepare'

const audio = new Audio(countdown);

class Timer {
  options = {
    workTime: 10,
    breakTime: 20
  }

  state: Ref<State> = ref('stopping')

  isBreakTime: boolean = false

  endDate: Ref<number> = ref(0)

  /** at seconds */
  timeLeft: Ref<number> = ref(this.options.workTime)

  private _interval: number | undefined = undefined
  private _timeout: number | undefined = undefined

  start () {
    this._clear()
    audio.play()
    this.state.value = 'prepare'
    this._timeout = setTimeout(() => {
      
      this.state.value = 'running'
      this.endDate.value = getEndDate(this.options.workTime)
      this._interval = setInterval(() => {
        if(this.timeLeft.value <= 0) {
          this.isBreakTime = !this.isBreakTime
          this.endDate.value = getEndDate(this.isBreakTime ? this.options.breakTime: this.options.workTime)
        } else if(this.timeLeft.value === 4) audio.play()
        this.timeLeft.value = Math.round((this.endDate.value - Date.now()) / 1000)
      }, 1000)
    }, 3000)
  }

  stop () {
    this.state.value = 'suspended'
    this._clear()
  }
  reset() {
    this.state.value = 'stopping'
    this._clear()
    this.isBreakTime = false
    this.endDate.value = 0
    this.timeLeft.value = this.options.workTime
  }
  private _clear(){
    clearInterval(this._interval)
    clearTimeout(this._timeout)
    audio.pause();
    audio.currentTime = 0;
  }
}

export default new Timer


function getEndDate(seconds: number) {
  return Date.now() + seconds * 1000
}