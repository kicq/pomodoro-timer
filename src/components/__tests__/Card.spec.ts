import { State } from '@/timer'
import { mount } from '@vue/test-utils'
import CardVue from '../Card.vue'

describe('Card.vue', () => {
  interface Props{
    seconds: number
    value: number
    isBreakTime: boolean
    state: State
  }

  const props: Props = {
    seconds: 0,
    value: 0,
    isBreakTime: false,
    state: 'running'
  }
  
  const wrapper = mount(CardVue, {props})
  it('test snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  
})