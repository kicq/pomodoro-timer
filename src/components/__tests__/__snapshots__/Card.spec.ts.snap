// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div class="card showGif">
  <div class="info">
    <section>
      <h2>Focus</h2>
      <!--v-if-->
    </section>
    <div class="circle-bar" radius="20"><svg id="svg" width="100" height="100" viewPort="0 0 50 50" version="1.1" xmlns="http://www.w3.org/2000/svg">
        <circle r="45" cx="50" cy="50" fill="transparent" stroke-width="10px" stroke-dasharray="314.1592653589793" stroke-dashoffset="0"></circle>
        <circle id="bar" r="45" cx="50" cy="50" fill="transparent" stroke-width="10px" stroke-dasharray="314.1592653589793" stroke-dashoffset="314.1592653589793px"></circle>
      </svg>
      <div class="legend" style="font-size: 24px;">00:00</div>
    </div>
  </div>
</div>
`;

exports[`Card.vue test snapshot 1`] = `
<div class="card showGif">
  <div class="info">
    <section>
      <h2>Focus</h2>
      <!--v-if-->
    </section>
    <div class="circle-bar" radius="20"><svg id="svg" width="100" height="100" viewPort="0 0 50 50" version="1.1" xmlns="http://www.w3.org/2000/svg">
        <circle r="45" cx="50" cy="50" fill="transparent" stroke-width="10px" stroke-dasharray="314.1592653589793" stroke-dashoffset="0"></circle>
        <circle id="bar" r="45" cx="50" cy="50" fill="transparent" stroke-width="10px" stroke-dasharray="314.1592653589793" stroke-dashoffset="314.1592653589793px"></circle>
      </svg>
      <div class="legend" style="font-size: 24px;">00:00</div>
    </div>
  </div>
</div>
`;
